src/python/pants/backend/go/goals/package_binary.py

```diff
@@ -39,23 +50,45 @@ class GoBinaryFieldSet(PackageFieldSet, RunFieldSet):
 @rule(desc="Package Go binary", level=LogLevel.DEBUG)
 async def package_go_binary(field_set: GoBinaryFieldSet) -> BuiltPackage:
     main_pkg = await Get(GoBinaryMainPackage, GoBinaryMainPackageRequest(field_set.main))
-    main_pkg_analysis = await Get(
-        FallibleFirstPartyPkgAnalysis, FirstPartyPkgAnalysisRequest(main_pkg.address)
-    )
-    analysis = main_pkg_analysis.analysis
-    if not analysis:
-        raise ValueError(
-            f"Unable to analyze main package `{main_pkg.address}` for go_binary target {field_set.address}: {main_pkg_analysis.stderr}"
+
+    if main_pkg.is_third_party:
+        import_path = cast(str, main_pkg.import_path)
+
+        go_mod_address = main_pkg.address.maybe_convert_to_target_generator()
+        go_mod_info = await Get(GoModInfo, GoModInfoRequest(go_mod_address))
+
+        analysis = await Get(
+            ThirdPartyPkgAnalysis,
+            ThirdPartyPkgAnalysisRequest(
+                import_path,
+                go_mod_info.digest,
+                go_mod_info.mod_path,
+            ),
+        )
+
+        package_name = analysis.name
```

Comment:
this has to happen inside the branch for single-party/third-party as I can't find a common type for First/ThirdPartyAnalysis.